#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
data_path = '/content/drive/MyDrive/Colab Notebooks/Data pandas/supermarket_sales.csv'
df = pd.read_csv(data_path)

Mounted at /content/drive


In [2]:
df

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920    33.4320   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190   649.2990   

          Date   Time      Payment    cogs  gross margin percentage  \
0     1/5/2019  13:08      Ewallet  522.83                 4.761905   
1     3/8/2019  10:29         Cash   76.40                 4.761905   
2     3/3/2019  13:23  Credit card  324.31                 4.761905   
3    1/27/2019  20:33      Ewallet  465.76                 4.761905   
4     2/8/2019  10:37      Ewallet  604.17                 4.761905   
..         ...    ...          ...     ...                      ...   
995  1/29/2019  13:46      Ewallet   40.35                 4.761905   
996   3/2/2019  17:16      Ewallet  973.80                 4.761905   
997   2/9/2019  13:22         Cash   31.84                 4.761905   
998  2/22/2019  15:33         Cash   65.82                 4.761905   
999  2/18/2019  13:28         Cash  618.38                 4.761905   

     gross income  Rating  
0         26.1415     9.1  
1          3.8200     9.6  
2         16.2155     7.4  
3         23.2880     8.4  
4         30.2085     5.3  
..            ...     ...  
995        2.0175     6.2  
996       48.6900     4.4  
997        1.5920     7.7  
998        3.2910     4.1  
999       30.9190     6.6  

[1000 rows x 17 columns]

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [13]:
df.groupby('Branch')['gross income'].sum().sort_values(ascending=False)

Branch
C    5265.1765
A    5057.1605
B    5057.0320
Name: gross income, dtype: float64

Філіал С має найбільший прибуток


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [15]:
df.groupby('City')['gross income'].sum().sort_values(ascending=False)

City
Naypyitaw    5265.1765
Yangon       5057.1605
Mandalay     5057.0320
Name: gross income, dtype: float64

Відповідь - Naypyitaw


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [22]:
invoices_by_category = pd.pivot_table(df, index = 'Branch', columns = 'Product line', values = 'Unit price', aggfunc="size")
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [24]:
invoices_by_category['Electronic accessories'].idxmax()

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Питання 1. Чи впливає поєднання типу клієнта (Member/Normal) та статі на сумарний прибуток, середній чек та середню кількість товарів у чеку?

In [26]:
customer_analysis = df.groupby(['Customer type', 'Gender']).agg({'gross income': 'sum', 'Total': 'mean', 'Quantity': 'mean'})
customer_analysis

gross income       Total  Quantity
Customer type Gender                                    
Member        Female     4197.4735  337.727753  5.716475
              Male       3622.6905  316.985419  5.387500
Normal        Female     3796.9515  332.233256  5.737500
              Male       3762.2535  305.047581  5.204633

Висновок. Члени програми лояльності-жінки приносять найбільший сумарний прибуток
Бызнесу необхыдно сконцентруватись на Member жінках для отримання гарного прибутку

Питання 2. Чи відрізняється популярність категорій товарів (Product line) у різних містах по сумарному прибутку та кількості покупок?

In [27]:
city_category_analysis = pd.pivot_table(df,index='City',columns='Product line',values=['Invoice ID', 'gross income'],aggfunc={'Invoice ID':'count', 'gross income':'sum'})
city_category_analysis

Invoice ID                                         \
Product line Electronic accessories Fashion accessories Food and beverages   
City                                                                         
Mandalay                         55                  62                 50   
Naypyitaw                        55                  65                 66   
Yangon                           60                  51                 58   

                                                                     \
Product line Health and beauty Home and lifestyle Sports and travel   
City                                                                  
Mandalay                    53                 50                62   
Naypyitaw                   52                 45                45   
Yangon                      47                 65                59   

                       gross income                                         \
Product line Electronic accessories Fashion accessories Food and beverages   
City                                                                         
Mandalay                   811.9735            781.5865           724.5185   
Naypyitaw                  903.2845           1026.6700          1131.7550   
Yangon                     872.2435            777.7385           817.2905   

                                                                     
Product line Health and beauty Home and lifestyle Sports and travel  
City                                                                 
Mandalay               951.460           835.6745          951.8190  
Naypyitaw              791.206           661.6930          750.5680  
Yangon                 599.893          1067.4855          922.5095

Mandalay
Найбільше покупок: Fashion accessories і Sports and travel (62 інвойси)
Найбільший прибуток: Sports and travel - 951.82

Naypyitaw
Найбільше покупок: Fashion accessories - 65 інвойсів
Найбільший прибуток: Food and beverages - 1131.76

Yangon
Найбільше покупок: Electronic accessories - 60 інвойсів
Найбільший прибуток: Home and lifestyle - 1067.49

Бізнес висновки


Mandalay:
Sports and travel приносить найбільший прибуток - варто розширювати асортимент цих товарів та акції на них.

Naypyitaw:
Food and beverages високий прибуток при великій кількості покупок - добре продаються, можна запускати акції “комбо-набори” для збільшення середнього чека.

Yangon:
Home and lifestyle приносить найбільший прибуток, хоча найбільше інвойсів – Electronic accessories =→ можливо, варто підвищити маржу на популярні товари або стимулювати покупки Home and lifestyle.

Питання 3. Чи залежить прибутковість і середній рейтинг від філіалу (Branch) та способу оплати (Payment)?

In [29]:
branch_payment_analysis = df.groupby(['Branch', 'Payment']).agg({'gross income':'sum','Total':'mean','Rating':'mean'})
branch_payment_analysis

gross income       Total    Rating
Branch Payment                                        
A      Cash            1608.6310  307.102282  6.938182
       Credit card     1575.9405  318.218755  6.985577
       Ewallet         1872.5890  312.098167  7.138889
B      Cash            1682.8315  321.267832  6.998182
       Credit card     1778.3265  342.613362  6.760550
       Ewallet         1595.8740  296.578354  6.698230
C      Cash            2051.7075  347.466593  6.973387
       Credit card     1444.1650  309.463929  7.291837
       Ewallet         1769.3040  350.522491  6.986792

Рекомендації для бізнес-рішень

Для Branch A варто просувати оплату через Ewallet, бо вона дає найвищий дохід і середній рейтинг вище 7.
Для Branch B – акцент на Credit card, для Branch C – на Cash.

Відстежувати задоволеність клієнтів (Rating) по способу оплати. Наприклад, у Branch B Ewallet показує найнижчий рейтинг, тож можна покращити сервіс при цьому способі оплати.